File: test/src/simple/xml/load/VersionTest.java
package simple.xml.load;

import junit.framework.TestCase;
import simple.xml.stream.Node;
import simple.xml.stream.NodeMap;
import simple.xml.Serializer;
import simple.xml.Attribute;
import simple.xml.Element;
import simple.xml.Root;

import java.lang.reflect.Constructor;
import java.util.Map;

public class VersionTest extends TestCase {

   private static final String VERSION_ATTRIBUTE = "version";        

   private static final String VERSION_1 =
   "<?xml version=\"1.0\"?>\n"+
   "<Example version='1'>\n"+
   "   <text>text value</text>  \n\r"+
   "</Example>";

   private static final String VERSION_2 =
   "<?xml version=\"1.0\"?>\n"+
   "<Example version='2'>\n"+
   "   <name>example name</name>  \n\r"+
   "   <value>text value</value> \n"+
   "   <entry name='example'>\n"+   
   "      <value>text value</value> \n"+   
   "   </entry>\n"+
   "</Example>";
   
   
   public interface Versionable {
      
      public int getVersion();
   }

   @Root(name="Example")
   private static abstract class Example implements Versionable {

      @Attribute(name="version")
      private int version;
      
      public int getVersion() {
         return version;
      }
      
      public abstract String getValue();   
   }
   
   private static class Example1 extends Example {

      @Element(name="text")
      private String text;           

      public String getValue() {
         return text;              
      }
   }

   private static class Example2 extends Example {          

      @Element(name="name")
      private String name;

      @Element(name="value")
      private String value;

      @Element(name="entry")
      private Entry entry;

      public String getValue() {
         return value;              
      }
   }

   private static class Entry {

      @Attribute(name="name")
      private String name;           

      @Element(name="value")
      private String value;              
   }

   public class VersionStrategy implements Strategy {

      private String version;           

      public Type getRoot(Class field, NodeMap root, Map map) throws Exception {
         Node version = root.get(VERSION_ATTRIBUTE);                       

         if(version != null){
            map.put(VERSION_ATTRIBUTE, version.getValue());
         }
         return getElement(field, root, map);
      }

      public Type getElement(Class field, NodeMap node, Map map) throws Exception {
         String value = field.getName() + map.get(VERSION_ATTRIBUTE);
     
         try {    
            Class type = Class.forName(value);
            
            return new SimpleType(type);
         } catch(ClassNotFoundException e) {
            return null;                 
         }            
      }         

      public void setRoot(Class field, Object value, NodeMap root, Map map) throws Exception {
         Class type = value.getClass();
         
         if(Versionable.class.isAssignableFrom(type)) {
            Versionable versionable = (Versionable)value;
            map.put(VERSION_ATTRIBUTE, String.valueOf(versionable.getVersion()));
         }              
         setElement(field, value, root, map);
      }              

      public void setElement(Class field, Object value, NodeMap node, Map map) throws Exception {
         node.put(VERSION_ATTRIBUTE, (String)map.get(VERSION_ATTRIBUTE));
      }
   }
   
   public static class SimpleType implements Type{
	   
	   private Class type;
	   
	   public SimpleType(Class type) {
		   this.type = type;
	   }
	   
	   public Object getInstance() throws Exception {
		   return getInstance(type);
	   }

       private Object getInstance(Class type) throws Exception {
		   Constructor method = type.getDeclaredConstructor();

		   if(!method.isAccessible()) {
		      method.setAccessible(true);              
		   }
		   return method.newInstance();   
	   }   

	   public Class getType() {
		  return type;
	   } 
   }
   
   public void testVersion1() throws Exception {    
      Strategy strategy = new VersionStrategy();           
      Serializer persister = new Persister(strategy);
      Example example = persister.read(Example.class, VERSION_1);
      
      assertTrue(example instanceof Example1);
      assertEquals(example.getValue(), "text value");
   }

   public void testVersion2() throws Exception {    
      Strategy strategy = new VersionStrategy();           
      Serializer persister = new Persister(strategy);
      Example example = persister.read(Example.class, VERSION_2);      
      
      assertTrue(example instanceof Example2);
      assertEquals(example.getValue(), "text value");
   }
}

File: port/src/main/Xml/VersionTest.cs
#region Using directives
using SimpleFramework.Xml.Strategy;
using SimpleFramework.Xml;
using System;
#endregion
namespace SimpleFramework.Xml.Core {
   public class VersionTest : TestCase {
      private const String VERSION_1 =
      "<?xml version=\"1.0\"?>\n"+
      "<Example version='1.0'>\n"+
      "   <text>text value</text>  \n\r"+
      "</Example>";
      private const String VERSION_2 =
      "<?xml version=\"1.0\"?>\n"+
      "<Example version='2.0'>\n"+
      "   <name>example name</name>  \n\r"+
      "   <value>text value</value> \n"+
      "   <entry name='example'>\n"+
      "      <value>text value</value> \n"+
      "   </entry>\n"+
      "   <ignore>ignore this element</ignore>\n"+
      "</Example>";
      public interface Versionable {
         public abstract double Version {
            get;
         }
         //public double GetVersion();
      [Root(Name="Example")]
      private static abstract class Example : Versionable {
         [Version]
         [Namespace(Prefix="prefix", Reference="http://www.domain.com/reference")]
         private double version;
         public double Version {
            get {
               return version;
            }
         }
         //public double GetVersion() {
         //   return version;
         //}
      }
      private static class Example1 : Example {
         [Element(Name="text")]
         private String text;
         public String Value {
            get {
               return text;
            }
         }
         //public String GetValue() {
         //   return text;
         //}
      private static class Example2 : Example {
         [Element(Name="name")]
         private String name;
         [Element(Name="value")]
         private String value;
         [Element(Name="entry")]
         private Entry entry;
         public String Value {
            get {
               return value;
            }
         }
         //public String GetValue() {
         //   return value;
         //}
      private static class Entry {
         [Attribute(Name="name")]
         private String name;
         [Element(Name="value")]
         private String value;
      }
      public static class SimpleType : Value{
         private Class type;
         public SimpleType(Class type) {
            this.type = type;
         }
         public int Length {
            get {
               return 0;
            }
         }
         //public int GetLength() {
         //   return 0;
         //}
            try {
               Constructor method = type.getDeclaredConstructor();
               if(!method.isAccessible()) {
                  method.setAccessible(true);
               }
               return method.newInstance();
            }catch(Exception e) {
               throw new RuntimeException(e);
            }
         }
          public Object Value {
             set {
             }
          }
          //public void SetValue(Object value) {
          //}
             return false;
          }
         public Class Type {
            get {
              return type;
            }
         }
         //public Class GetType() {
         //  return type;
         //}
      public void testVersion1() {
         Serializer persister = new Persister();
         Example example = persister.read(Example1.class, VERSION_1);
         assertTrue(example instanceof Example1);
         assertEquals(example.Value, "text value");
         assertEquals(example.version, 1.0);
         persister.write(example, System.out);
         assertEquals(example.Value, "text value");
         assertEquals(example.version, 1.0);
      }
      public void testVersion2() {
         Serializer persister = new Persister();
         Example example = persister.read(Example2.class, VERSION_2);
         assertTrue(example instanceof Example2);
         assertEquals(example.Value, "text value");
         assertEquals(example.version, 2.0);
         persister.write(example, System.out);
         assertEquals(example.Value, "text value");
         assertEquals(example.version, 2.0);
      }
   }
}
